### PR TITLE
feat: Add accessibility scenarios for Fluent UI v9 components #3

### PR DESCRIPTION
**Pull request by adamsamec**
#### Description of changes

This PR adds accessibility scenarios for testing of the Checkbox, Input, RadioGroup, Spinner, Switch, TabList and Textarea Fluent UI v9 components using a screen reader, and modifies the scenarios for some other existing  components mainly so that they use Fluent UI components instead of native HTML elements.
[The list of scenarios for testing can be found here](https://fluentuipr.z22.web.core.windows.net/pull/23334/public-docsite-v9/storybook/index.html?path=/docs/concepts-developer-accessibility-stories-list-of-scenarios--page).

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `181.029 kB`<br />`50.535 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            |   `33.95 kB`<br />`11.044 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/287cf6769bda0ca0ec7cd29b01a4a4481a488d92'>287cf6769bda0ca0ec7cd29b01a4a4481a488d92</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/Avatar-mount_base.html">1176</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/Avatar-mount.html">1249</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/Button-mount_base.html">754</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/Button-mount.html">731</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/FluentProvider-mount_base.html">2277</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/FluentProvider-mount.html">2209</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/FluentProviderWithTheme-mount_base.html">383</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/FluentProviderWithTheme-mount.html">390</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">346</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">348</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">412</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">434</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/MakeStyles-mount_base.html">1957</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23334/perf-test-react-components/MakeStyles-mount.html">1888</a></td>
            <td>50000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by bsunderhus** [comment]
@adamsamec , should we merge this PR? it's been approved for a while now.

**Comment by adamsamec** [comment]
Please, don't merge it yet. It's still work in progress.

File: packages/react-components/react-components/src/AccessibilityScenarios/Button.stories.tsx [hunk 1]
```diff
@@ -71,38 +71,30 @@ export const MessengerButtonsAccessibilityScenario: React.FunctionComponent = ()
...
+        name="message"
+        rows={3}
+        cols={50}
+        placeholder="Enter message here...."
+        aria-label="Message"
+        onChange={onMessageTextareaChange}
+        value={message}
+        style={messageStyle}
+      />
+      <Button disabledFocusable={sendButtonDisabled} onClick={onSendButtonClick}>
```
Comment (smhigley) [inline packages/react-components/react-components/src/AccessibilityScenarios/Button.stories.tsx:90]: This is an interesting thing to test, but it feels pretty weird to have the send/delete buttons be disabled and focusable. If this is just used to gather feedback from users, it seems fine. But if we plan on having this on our public docs, we should probably update it so the send button at least is always enabled.

File: packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx [hunk 2]
```diff
@@ -108,20 +130,18 @@ const RegistrationFormInputsAccessibility = () => {
...
       return;
     }
     const firstErrorName = Object.keys(errors)[0] as keyof FormInputs;
     const firstErrorField = document.getElementById(firstErrorName);
-
-    setTimeout(() => formValidation.notifyFormFieldError(firstErrorName), 200);
-
     if (firstErrorField) {
-      firstErrorField.focus();
+      setTimeout(() => firstErrorField.focus(), 500);
```
Comment (smhigley) [inline packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx:144]: what's the purpose of the half-second timeout?

File: packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx [hunk 3]
```diff
@@ -305,24 +318,106 @@ const RegistrationFormInputsAccessibility = () => {
...
+
+          <Label htmlFor="email">E-mail:</Label>
+          <Controller
+            name="email"
+            control={control}
+            as={
+              <Input
+                type="text"
+                id="email"
+                disabled={!isSendNewsletter}
```
Comment (smhigley) [inline packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx:371]: Similar comment as before, this only applies if we're exposing these scenarios on our docs site for authors:

as a general usability best practice, I think it'd be better to show/hide this field based on `!isSendNewsletter` rather than toggling disabled state.

File: packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+    setIsAlerting(false);
+    setTimeout(() => setIsAlerting(true), 200);
+  }, [setIsAlerting]);
+
+  React.useEffect(() => {
+    formValidation.subscribe(id, alert);
+    return () => formValidation.unsubscribe(id, alert);
+  }, [formValidation, alert, id]);
+  return (
+    <div role={isAlerting ? 'alert' : undefined} style={{ color: isAlerting ? 'red' : 'green' }} id={`${id}Errors`}>
```
Comment (smhigley) [inline packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx:55]: I believe dynamically changing the role is going to be less robust than dynamically inserting the element into the DOM with a consistent `role="alert"`, and using a different element without an alert role for the non-alert message

File: packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+interface ValidationMessageProps {
+  id: string;
+  formValidation: FormValidation;
+}
+const ValidationMessage: React.FC<ValidationMessageProps> = ({ id, formValidation, children }) => {
+  const [isAlerting, setIsAlerting] = React.useState(true);
+
+  const alert = React.useCallback(() => {
+    setIsAlerting(false);
+    setTimeout(() => setIsAlerting(true), 200);
```
Comment (smhigley) [inline packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx:47]: This timeout and focus behavior causes focus to get sent back to the first field in the form that has an error, even if someone is working on an entirely different field and tabs away. In more detail: if I leave the first field in an error state to come back to later and move on, tabbing away from e.g. the 3rd field and triggering validation there will randomly send focus back to the first field.

I believe a better approach would be to only shift focus when a user activates the submit button.

**Review by smhigley — COMMENTED** [review]
I like the new scenarios! Left a couple comments, the only important one was the note on the focus shifting.

File: packages/react-components/react-components/src/AccessibilityScenarios/Button.stories.tsx (hunk 1 above)
Comment (adamsamec) [reply packages/react-components/react-components/src/AccessibilityScenarios/Button.stories.tsx]: It's not so easy to find a case when disabledFocusable would make sense, but you are right here... I have made the "Send" button enabled at all times. Do you think I could leave the "Delete" button disabledFocusable at least?

File: packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx (hunk 3 above)
Comment (adamsamec) [reply packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx:371]: I agree in the principle. Though as the purpose of these scenarios is at the first place to present to the testers all the different variants and types of the components in question, I have been a bit struggling to find a meaningful example of all these variants and types. Please, do you have a better idea for a disabled input to demonstrate in this scenario?

File: packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx (hunk 5 above)
Comment (adamsamec) [reply packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx:47]: The focus should indeed move only when activating the "Send" button. When I tab through edit fields, the focus does not move to the first error field in my case. Could you please provide exact repro steps?

File: packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx (hunk 2 above)
Comment (adamsamec) [reply packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx]: I cannot recall the reason for it now. However, I have removed it and looks it works okay :-).

File: packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx (hunk 4 above)
Comment (adamsamec) [reply packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx]: Okay, I have refactored it to insert a whole different DOM element instead of changing just role, though it looks like it works the same as before.

File: packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx (hunk 5 above)
Comment (smhigley) [reply packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx:47]: First submit the form once without making any changes. Next navigate to the second textarea, and add a few characters (so the error message changes), and tab away. Your focus is then programmatically sent to the first textarea after a half-second delay.

The bug looks like it's in the `useEffect` on line 119, since after the form is submitted once, it runs again every time `errors` changes.

This same bug exists on the Input story.

File: packages/react-components/react-components/src/AccessibilityScenarios/Button.stories.tsx (hunk 1 above)
Comment (smhigley) [reply packages/react-components/react-components/src/AccessibilityScenarios/Button.stories.tsx]: How about making it `tabindex="-1"` when disabled and then that way if it becomes disabled because someone clicks it, keyboard focus doesn't shift, but it's still not in the tab order (unless enabled). I think that'd be a pretty good real-world sue case.

File: packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx (hunk 3 above)
Comment (smhigley) [reply packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx:379]: I think the current use case is good for testing, and it's probably better to leave them as-is for now and run tests to get feedback sooner than later. Then maybe we can revisit this if we want to repurpose the examples for showing authors.

File: packages/react-components/react-components/src/AccessibilityScenarios/Button.stories.tsx (hunk 1 above)
Comment (adamsamec) [reply packages/react-components/react-components/src/AccessibilityScenarios/Button.stories.tsx]: That sounds like a good idea. So should I make the following changes?

* Leave the "Send" button always enabled.
* Make the "Delete" button disabled when activated, but set tabindex="-1" so that it remains in focus after becoming disabled.
* If there is no more steps for the "Decrease font size" or "Increase font size" buttons, make them disabledFocusable instead of disabled.
*

File: packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx (hunk 3 above)
Comment (adamsamec) [reply packages/react-components/react-components/src/AccessibilityScenarios/Input.stories.tsx:379]: Okay, anyway, do you please have a good idea for a disabled input example? It does not necessary hav to be part of this scenario. I could create a new scenario for that.

File: packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx (hunk 5 above)
Comment (adamsamec) [reply packages/react-components/react-components/src/AccessibilityScenarios/Textarea.stories.tsx:47]: Thanks for the cue. Looks like I have fixed the bug, but I had to reintroduce the half second timeout for focusing the first error field upon form submission. Without the timeout, the focus sometimes stood at the submit button after submission.

File: packages/react-components/react-components/src/AccessibilityScenarios/Button.stories.tsx (hunk 1 above)
Comment (adamsamec) [reply packages/react-components/react-components/src/AccessibilityScenarios/Button.stories.tsx]: So I've tried setting tabindex="-1" on the "Delete" button after clicking it, where clicking it also makes it disabled, but I cannot set focus programatically on the button afterwards, because it is not possible to set focus programatically on a disabled element. Therefore unfortunately what you suggested is not possible to achieve.

File: apps/public-docsite-v9/.storybook/main.utils.js [hunk 6]
```diff
@@ -24,7 +24,7 @@ function getVnextStories() {
     .filter(pkgName => pkgName.startsWith('@fluentui/') && !excludedDependencies.includes(pkgName))
     .map(pkgName => {
       const name = pkgName.replace('@fluentui/', '');
-      const storiesGlob = '**/@(index.stories.@(ts|tsx)|*.stories.mdx)';
+      const storiesGlob = '**/@(index.stories.@(ts|tsx)|*.stories.mdx|*.stories.tsx)';
```
Comment (micahgodbolt) [inline apps/public-docsite-v9/.storybook/main.utils.js:27]: we moved away from the *.stories.tsx glob to remove all of the invalid files that storybook tries to load. The new pattern is to use /folderName/index.stories.tsx, and then index re-exports all of the foo.stories.tsx files.

Comment (micahgodbolt) [reply apps/public-docsite-v9/.storybook/main.utils.js:27]:
```suggestion
       const storiesGlob = '**/@(index.stories.@(ts|tsx)|*.stories.mdx)';
```

File: packages/react-components/react-persona/MIGRATION.md [hunk 7]
```diff
@@ -13,6 +13,7 @@ Here's how the API of v8's `Persona` compares to the one from v9's `Persona` com
 - ```
 
   ```
+
```
Comment (sopranopillow) [inline packages/react-components/react-persona/MIGRATION.md:16]: Could you revert this change? you can use `--no-verify` when commiting so it doesn't change it back.

**Review by sopranopillow — APPROVED** [review]
